### DISCUSSION TITLE
Video bitrate vs. h264 keyframes

### DISCUSSION THREAD

**Discussion by frogblueDFi** [baresip/baresip#2707] [Q&A]
Hi to all!

When trying to improve the video performance of our SIP app based on baresip we are playing around with video size, framerates and video bitrate.
Especially when changing the default video.bitrate config we observe a better latency together with loss of video quality when lowering the bitrate (about 250000 for CIF resolution).
Increasing the bitrate (for CIF resolution) to 1000000 we see a worse latency and when increasing the bitrate even more it comes to a situation, where we don't get a video at all or very very late.
Reading the logs shows that no h264 keyframes are received 

> avcodec: using H.264 encoder 'libx264' -- libx264 H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
> avcodec: using H.264 decoder 'h264' -- H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
> avcodec: using H.265 decoder 'hevc' -- HEVC (High Efficiency Video Coding)
> avcodec: enable hwaccel using 'videotoolbox'
> avcodec: decode: using hardware pixel format 'videotoolbox_vld'
> 
> video tx pipeline: frogcapture ---> avfilter ---> H264
> video rx pipeline:  frogvideo <--- H264
> 
> avcodec: decoder stats (keyframes:0, lost_fragments:0)

What is the relation between bitrate and h264 encoding and latency?

Thanx for any hint!
Dietmar

- FFmpeg with libx264 and VideoToolbox hardware encoding acceleration for iOS/macOS.
- Baresip as static lib.
- Config:
# SIP
sip_listen		192.168.132.18:5060
sip_certificate	
sip_cafile		
sip_capath		
sip_transports		UDP,TCP,TLS,WS,WSS
sip_trans_def	UDP
sip_verify_server			no
sip_tos	160

# Call
call_local_timeout	120
call_max_calls		1
call_hold_other_calls	yes

# Audio
audio_path		
audio_player		coreaudio,
audio_source		coreaudio,
audio_alert		coreaudio,
auplay_srate		16000
ausrc_srate		16000
auplay_channels		1
ausrc_channels		1
audio_txmode		poll
audio_level		no
ausrc_format		S16LE
auplay_format		S16LE
auenc_format		S16LE
audec_format		S16LE
audio_buffer		20-160		# ms
audio_buffer_mode	fixed		# fixed, adaptive
audio_silence		-35.0		# in [dB]
audio_telev_pt		101

# Video
video_source		frogcapture,Logitech Webcam C925e
#video_source		avformat,rtmp://127.0.0.1/app/foo
video_display		frogvideo,
video_size		"640x480"
video_bitrate		2000000
video_fps		15.00
video_fullscreen	no
videnc_format		nv12

# AVT
rtp_tos			184
rtp_video_tos		136
rtp_ports		1024-49152
rtp_bandwidth		512-4096
audio_jitter_buffer_type	fixed
audio_jitter_buffer_delay	5-10
video_jitter_buffer_type	fixed
video_jitter_buffer_delay	5-50
rtp_stats		yes
rtp_timeout		0 # in seconds

# Network
net_interface		
net_af			ipv4

**Comment by sreimers** [comment]
You have two more config options to optimize the latency, with higher bitrates you can try:

```
video_sendrate = 2 x video_bitrate
video_burst_bit = video_sendrate / 10
```

But the values depend on the available uplink bandwidth.

**Comment by frogblueDFi** [comment]
Thank you very much for the quick response.
The video.send_bitrate and video.burst_bits have indeed an influence on the latency (better).
At least with CIF resolution , 12.0 fps and bitrate 250000.

For VGA/15.0 fps/500000 and HD/30fps/500000 latency is also good.

Default bitrate 1000000 always cause bad latency.
Bitrates above 1000000 cause very high CPU load and are not usable with the settings for send_bitrate and burst_bits.

Are there any rules/suggestions/heuristics for setting the best combination of all this?

Note: I am using direct calls within reliable WLAN for testing. linphone app for reference. Facetime is the benchmark :-)

Thank you again
Dietmar

**Comment by frogblueDFi** [comment]
Some progress on this:

As I still had bad video performance (and almost a freezing UI due to high CPU load) for anything above CIF resolution, I was wondering if VideoToolbox hardware acceleration really takes place even I could see log entry: "avcodec: enable hwaccel using 'videotoolbox'"
Compiling FFmpeg in debug mode and setting breakpoints in module videotoolboxenc.c showed that VideoToolbox is not used for encoding.
Next step was to explicitly enable h264 VideoToolbox encoder via baresip config:

```
avcodec_hwaccel videotoolbox
avcodec_h264enc h264_videotoolbox
avcodec_h264dec h264
avcodec_keyint 10
```
Now breakpoints in videotoolboxenc.c are hit! Guess iff FFmpeg is build with libx264 support, VideoToolBox hardware encoder will not be used.

But now: Latency is worse compared to software h264 encoding. For larger video size video even freezes after a while!
If video freezes I see log entries "avcodec: ignoring fragment (nal=5)“ on the receiving part.

Anybody any idea how to correctly use VideoToolBox on iOS/macOS?
Or how to set baresip parameters for avcodec in detail?

Thanx in advance
Dietmar

**Comment by frogblueDFi** [comment]
No progress on this ☹️

Still the same behaviour both for macOS and iOS. Using VideoToolbox hardware acceleration does not really improve video latency.
At least CPU load is acceptable for higher resolutions.

But even worse:
When sending video to another instance of my own baresip implementation I still observe that no video is shown or freezes after a while due to missing h264 keyframes.

Sending video to e.g. Linphone is ok.
Also receiving high resolution video with high framerate (e.g. from Linphone) is ok.

What is the influence of video_bitrate, video_sendrate and video_burst_bit on this?
Guess it make no sense for me to play with avcodec options. This has been done by others with more knowledge.

Someone out there using baresip video for macOS/iOS?

I am still using re/baresip 3.2.0 and FFmpeg n6.0.
For completeness I add the configure and cmake calls for FFmpeg and baresip.

Thanx again in advance for any hints!

```
./configure --prefix=/<path>/baresip/macOS/build/arm64 --target-os=darwin --arch= arm64 --cc=xcrun -sdk macosx clang --cxx=xcrun -sdk macosx clang++ --enable-cross-compile --extra-cflags=-arch arm64 -mmacos-version-min=13.0 -I/<path>/baresip/macOS/build/arm64/include --extra-ldflags=-arch arm64 -L/<path>/baresip/macOS/ffmpeg/../lib/arm64 -lx264 --enable-gpl --enable-libx264 --disable-programs --disable-doc --disable-appkit --disable-audiotoolbox --disable-sdl2 --disable-securetransport --disable-libxcb --enable-debug

cmake -B build --install-prefix=/<path>/baresip/macOS/build/arm64 -DCMAKE_OSX_ARCHITECTURES=arm64 -DCMAKE_OSX_DEPLOYMENT_TARGET=13.0 -DSTATIC=ON -DUSE_OPENSSL=ON -DOPENSSL_ROOT_DIR=/<path>/baresip/macOS/build/arm64 -DSPANDSP_INCLUDE_DIR=/<path>/baresip/macOS/build/arm64/include -DSPANDSP_LIBRARY=/<path>/baresip/macOS/build/arm64/lib/libspandsp.a -DFFMPEG_avcodec_LIBRARY=/<path>/baresip/macOS/build/arm64/lib/libavcodec.a -DFFMPEG_avcodec_INCLUDE_DIR=/<path>/baresip/macOS/build/arm64/include -DFFMPEG_avdevice_LIBRARY=/<path>/baresip/macOS/build/arm64/lib/libavdevice.a -DFFMPEG_avdevice_INCLUDE_DIR=/<path>/baresip/macOS/build/arm64/include -DFFMPEG_avfilter_LIBRARY=/<path>/baresip/macOS/build/arm64/lib/libavfilter.a -DFFMPEG_avfilter_INCLUDE_DIR=/<path>/baresip/macOS/build/arm64/include -DFFMPEG_avformat_LIBRARY=/<path>/baresip/macOS/build/arm64/lib/libavformat.a -DFFMPEG_avformat_INCLUDE_DIR=/<path>/baresip/macOS/build/arm64/include -DFFMPEG_avutil_LIBRARY=/<path>/baresip/macOS/build/arm64/lib/libavutil.a -DFFMPEG_avutil_INCLUDE_DIR=/<path>/baresip/macOS/build/arm64/include -DFFMPEG_swscale_LIBRARY=/<path>/baresip/macOS/build/arm64/lib/libswscale.a -DFFMPEG_swscale_INCLUDE_DIR=/<path>/baresip/macOS/build/arm64/include -DFFMPEG_swresample_LIBRARY=/<path>/baresip/macOS/build/arm64/lib/libswresample.a -DFFMPEG_swresample_INCLUDE_DIR=/xxx/baresip/macOS/build/arm64/include -DMODULES=auconv
```

**Comment by sreimers** [comment]
If no packet lost is involved, how does your `video_jitter_buffer` settings look like? Maybe it's to small for keeping/reorder all packets on receive side?

**Comment by frogblueDFi** [comment]
No packets are lost!

  ```
baresipConfig->avt.video.jbtype  = jbuf_type::JBUF_FIXED;
baresipConfig->avt.video.jbuf_del.min = 5;
baresipConfig->avt.video.jbuf_del.max = 50;
```

**Reply by sreimers** [reply]
50 max. packets is very low. The max value is for packets, the min. is frames (5 frames with 30 fps are ~166ms min. delay). you can try:
```
baresipConfig->avt.video.jbuf_del.min = 2;
baresipConfig->avt.video.jbuf_del.max = 500;
```

**Comment by frogblueDFi** [comment]
![icon_128x128](https://github.com/baresip/baresip/assets/129487688/b086a6f5-ac50-4dae-8e1d-f4200c83d3bf)

Wow! Thank you very very much! That fix it!
How could a stupid user as me guess this values?

Now I can try to increase framerate and latency again.
Are there any other typical config values beside video.bitrate, video.send_bitrate and video.burst_bits?

Still trying to achieve same performance as FaceTime 😎

Dietmar
